File: code/09-string-matching/python/horspool_demo.py
from string_matching import build_horspool_shift, horspool_search

if __name__ == "__main__":
    tiger_shift, _ = build_horspool_shift("TIGER")
    rational_shift, _ = build_horspool_shift("RATIONAL")
    print("shift_I:", tiger_shift["I"])
    print("shift_A:", rational_shift["A"])
    text, pattern = "acebbceeaabceedb", "eeaab"
    matches = horspool_search(text, pattern)
    print("matches:", ",".join(str(s) for s in matches))

File: code/09-string-matching/java/HorspoolDemo.java
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public final class HorspoolDemo {
    public static void main(String[] args) {
        int[] fallback = new int[1];
        Map<Character,Integer> tigerShift = StringMatchers.buildHorspoolShift("TIGER", fallback);
        Map<Character,Integer> rationalShift = StringMatchers.buildHorspoolShift("RATIONAL", fallback);
        System.out.println("shift_I: " + tigerShift.get('I'));
        System.out.println("shift_A: " + rationalShift.get('A'));
        String text = "acebbceeaabceedb", pattern = "eeaab";
        List<Integer> matches = StringMatchers.horspoolSearch(text, pattern);
        System.out.println("matches: " + matches.stream().map(String::valueOf).collect(Collectors.joining(",")));
    }
}
